elki-core-math/src/main/java/elki/math/statistics/tests/mcde/MCDETest.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * This file is part of ELKI:
+ * Environment for Developing KDD-Applications Supported by Index-Structures
+ *
+ * Copyright (C) 2019
+ * ELKI Development Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package elki.math.statistics.tests.mcde;
+
+import elki.math.MathUtil;
+import elki.utilities.datastructures.arraylike.NumberArrayAdapter;
+import elki.utilities.datastructures.arrays.IntegerArrayQuickSort;
+
+/**
+ * Abstract class for statistical tests for MCDEDependenceMeasure.
+ * See MWPTest for an example. Implementation should contain an appropriate index
+ * structure to efficiently compute statistical test and the test itself.
+ *
+ * @param <R> RankStruct or Subclass of RankStruct
+ *
+ * @author Alan Mazankiewicz
+ * @author Edouard Fouché
+ */
+
+public abstract class MCDETest<R extends MCDETest.RankStruct> {
+
+  /**
+   * Structure to hold return values in index creation for MCDEDependenceEstimate
+   */
+  public class RankStruct {
```

Comment:
should be a `static` inner class (saves some memory), and should have JavaDoc on attributes and methods.